<b><center>UNCLASSIFIED</center></b>

<img src="./GRAPHICS/Banner.png" width="110%">

# Lesson 6: Getting Tabular Data

<hr>
## Table of Contents

* [6.1. Objectives](#Objectives)<br>
* [6.2. Overview](#Overview)<br>
* [6.3. Review](#Review)<br>
   * [6.3.1. Entering Data From the Prompt](#EnteringData)<br>
   * [6.3.2. Loading a Local Data File](#LoadingLocal)<br>
* [6.4. Lesson: Getting Tabular Data](#Lesson)<br>
   * [6.4.1. Getting Data with the CSV Library](#CSV)<br>
   * [6.4.2. The Glob Library: How to Read A LOT of Data into a Python Script](#Glob)<br>
* [6.5. Guided Exercise: Data Exploration and Engineering](#GuidedExercise)<br>
* [6.6. Practical Exercises](#PracticalExercise)<br>
   * [6.6.1. Practical Exercise 1: Create a Function to Read a File](#PE1)<br>
   * [6.6.2. Practical Exercise 2: Clean a CSV File](#PE2)<br>
   * [6.6.3. Practical Exercise 3: Read in a CSV File Using the Glob Library](#PE3)<br>
   * [6.6.4. Practical Exercise 4: Make a Dictionary](#PE4)<br>
   * [6.6.5. Practical Exercise 5: Read a Sample Dataset](#PE5)<br>
* [6.7. Administrative Notes](#Administrative)<br>

***
<a id='Objectives'></a>
## 6.1. Objectives
Using conditionals, loops, Python dictionaries, local data, the CSV Library, and the Glob Library, students will be able to:

* Examine the implications of using computation to solve a problem
    * Discuss best practices for using computation to solve a problem
    * Suggest types of problems that can be solved through computation
    * Show how computation can solve a problem
<br><br>    
* Recognize key computer science concepts
    * Recognize how queries operate
<br><br>    
* Demonstrate the ability to build basic scripts using Python scripting language
    * Use various data types and structures in Python scripting
    * Collect data using Python scripting
    * Extract data using Python scripting
    * Develop advanced data structures using Python scripting

<hr>
<a id='Overview'></a>
## 6.2. Overview
The following lesson is divided into three main parts:
* Lesson: Working with Data and the Glob Library
* Guided Exercise: Data Exploration and Engineering
* Practical Exercises

<b><font color="brick">Instructor Guidance</font>: Refer back to Lesson 1 and relate the four steps of problem-solving using Computational Thinking (Decomposition, Pattern Recognition, Abstraction, & Algorithm Design) to lessons, exercises, examples, student questions/comments, etc., as appropriate throughout this lesson.</b>

<hr>
<a id='Review'></a>
## 6.3. Review

<hr>
<a id='EnteringData'></a>
### 6.3.1. Entering Data From the Prompt
<hr>
#### References:
* [Python 3.5 input()](https://docs.python.org/3/library/functions.html#input)
<hr>

<b><font color="brick">Instructor Guidance: </font> Ask the students for the syntax of the input statement.</b> 

One of the simplest ways to get data into a Python script is an `input()` statement. Remember that the `input()` statement on its own doesn't do much for us, we have to assign it to a variable to be able to use that input. 

In [1]:
name = input('Type your name! ')

print('Hello,', name + '! How are you today?')  

Type your name! bob
Hello, bob! How are you today?


<hr>
<a id='LoadingLocal'></a>
### 6.3.2. Loading a Local Data File

#### References:
* [Python 3.5 open()](https://docs.python.org/3.5/library/functions.html#open)
* [Python Tutorial - Reading and Writing Files](https://docs.python.org/2/tutorial/inputoutput.html#reading-and-writing-files)
<hr>

<b><font color="brick">Instructor Guidance: </font> Ask the students for the basic file loading algorithm. Make sure they understand that we need to first open the file, read in the contents and then close the file.</b>

The most basic way to read or write a file is using the `.read()` and `.write()` methods. We use the <b><font color="green">with</font></b> keyword to store the file in a temporary variable. Then we indent our code just like with conditionals and loops, and the file will close automatically when the indentation ends. 

* To read a file, the mode is `'r'` (default).
* To write a file, the mode is `'w'`.

```python
with open(path + filename, mode = 'r') as my_file:     
    data = my_file.read()
```

Put it all together by reading in a file that we will be using with today's lesson. 

In [2]:
with open('data/animal_info.csv') as f:
    animals = f.read().lower()

animals

'name,weight,diet,average lifespan\nblue whale,300000,krill,95\ngiraffe,2600,shrubbery,25\nmoose,1300,shrubbery,20\ngiant panda,220,bamboo,20\nred panda,14,bamboo,14\ngreat white shark,2000,fish,70\nocean sunfish,2200,jellyfish,10\nhippopotamus,3800,grass,45\nlion,400,gazelle,12\n'

Then we can use a membership test to check if a string that we input is in the data.

In [3]:
keyword = input('Food that an animal might eat: ')

if keyword.lower() in animals:
    print('It looks like an animal in the database eats', keyword)
    
else:
    print('No animal in the database eats', keyword)

Food that an animal might eat: grass
It looks like an animal in the database eats grass


As you can see from the file extension, we just read in a CSV file. Let's dig into this new file type's library and get some practice with this common file type.

<hr>
<a id='Lesson'></a>
## 6.4. Lesson: Getting Tabular Data 
<img src="./GRAPHICS/csvReader.png" width="30%"><br>
<a id='CSV'></a>
### 6.4.1. Getting Data with the CSV Library
<hr>
#### References:
* [Python 3.4: CSV File Reading and Writing](https://docs.python.org/3.4/library/csv.html)
<hr>


The comma separated values (CSV) format is a common import and export format for spreadsheets, databases, and programming languages. CSVs are for the transfer of data between systems and are readable by many applications and programs, including Excel. The CSV Library provides functions to both read and write CSV files. Below are a couple of terms that may be unfamiliar but are helpful to understand.

* `delimiter` is a character or string that separates individual pieces of data.
* `quotechar` wraps data that contains your delimiter character, allowing it to be read as one piece of data.

Whenever we work with libraries in Python, we need to import them into our script. We only have to do this once inside the current kernel, but if we restart the kernel we need to run the import statement again.

In [4]:
import csv

Now that we've imported the library, we can use `help(csv)` to explore all of its features.

In [5]:
help(csv)

Help on module csv:

NAME
    csv - CSV parsing and writing.

DESCRIPTION
    This module provides classes that assist in the reading and writing
    of Comma Separated Value (CSV) files, and implements the interface
    described by PEP 305.  Although many CSV files are simple to parse,
    the format is not formally defined by a stable specification and
    is subtle enough that parsing lines of a CSV file with something
    like line.split(",") is bound to fail.  The module supports three
    basic APIs: reading, writing, and registration of dialects.
    
    
    DIALECT REGISTRATION:
    
    Readers and writers support a dialect argument, which is a convenient
    handle on a group of settings.  When the dialect argument is a string,
    it identifies one of the dialects previously registered with the module.
    If it is a class or instance, the attributes of the argument are used as
    the settings for the reader or writer:
    
        class excel:
            delimiter = ',

<hr>
#### 6.4.1.1. Reading a Basic CSV File with `csv.reader()`

`csv.reader()` returns an object that all the data from each row in a CSV file. To retain the data, it must be appended to a list for you to manipulate later in the code. These objects are good for storing data but are not easy to work with. Because of this limitation, programmers typically push the data from the `.reader()` object into a standard Python list.

The code below, reads in a new CSV file stored on your computer. The variable `animal` stores the object created by `csv.reader()`. We then use a `for` loop to step through each row of data and print it to the screen.

In [6]:
with open('data/animal/animal_info_land.csv') as csvfile:
    
    animal = csv.reader(csvfile)
    
    for row in animal:
        print(row)

['Name', 'Weight', 'Diet', 'Average Lifespan']
['Giraffe', '2600', 'Shrubbery', '25']
['Moose', '1300', 'Shrubbery', '20']
['Giant Panda', '220', 'Bamboo', '20']
['Red Panda', '14', 'Bamboo', '14']
['Hippopotamus', '3800', 'Grass', '45']
['Lion', '400', 'Gazelle', '12']
['Tortoise', '100', 'Grass', '100']


Can you see the tabular structure? If you're curious what the `.reader()` object looks like, run the code below.

In [7]:
animal

Remember that printing data is not the same as storing it in memory. In order to save the information for further use, we need to append each row to a list.

In [8]:
land_animals = []

with open('data/animal/animal_info_land.csv') as csvfile:
    
    animal = csv.reader(csvfile)
    
    for row in animal:
        land_animals.append(row)

land_animals

[['Name', 'Weight', 'Diet', 'Average Lifespan'],
 ['Giraffe', '2600', 'Shrubbery', '25'],
 ['Moose', '1300', 'Shrubbery', '20'],
 ['Giant Panda', '220', 'Bamboo', '20'],
 ['Red Panda', '14', 'Bamboo', '14'],
 ['Hippopotamus', '3800', 'Grass', '45'],
 ['Lion', '400', 'Gazelle', '12'],
 ['Tortoise', '100', 'Grass', '100']]

1. What is the data type of each row?
2. What are all of the data types in each item within each row?
3. Does this structure make sense as a way to store information from a file? Why or why not?

<hr>
#### 6.4.1.2. Writing to a CSV File
Writer objects can take data and write it to a file in CSV format. We create them like this:

```python
csv.writer(csvfile)
```

Before we write our data out, let's change a piece of it. We do this for two reasons: 1) this will show that we've actually written the data we have, not just copied an old file, and 2) our new data value will have our delimiter character (`','`) in it, so this will demonstrate the use of quotechars. Here, we'll edit the third item in the last column of our table.

In [9]:
land_animals[-1][2] = 'Grass, Bananas'
land_animals

[['Name', 'Weight', 'Diet', 'Average Lifespan'],
 ['Giraffe', '2600', 'Shrubbery', '25'],
 ['Moose', '1300', 'Shrubbery', '20'],
 ['Giant Panda', '220', 'Bamboo', '20'],
 ['Red Panda', '14', 'Bamboo', '14'],
 ['Hippopotamus', '3800', 'Grass', '45'],
 ['Lion', '400', 'Gazelle', '12'],
 ['Tortoise', '100', 'Grass, Bananas', '100']]

Writing with the CSV Library is very similar to using the standard `.write()` method. The syntax is slightly different, because we first need to create a `.writer()` object before we can write any data. The `.writer()` object takes the file stream as an argument. To actually write data, we call the object and its method, `.writerows()`, and pass in a list of lists.

NOTE: The `newline=''` argument is important for writing CSVs that we intend to open in Excel. Try taking it out and see how the output changes.

In [10]:
with open('data/animal/animal_output.csv', mode='w', newline='') as csvfile:
    
    writer = csv.writer(csvfile)
    
    writer.writerows(land_animals) 

Can you find the file we just created on your computer? Where is this file located? You should be able to find this file in your file system and open it, but we'll just read the file back in to check that it worked.

In [11]:
animals_list = []

with open('data/animal/animal_output.csv') as csvfile:
    
    animal = csv.reader(csvfile)
    
    for row in animal:
        animals_list.append(row)

animals_list

[['Name', 'Weight', 'Diet', 'Average Lifespan'],
 ['Giraffe', '2600', 'Shrubbery', '25'],
 ['Moose', '1300', 'Shrubbery', '20'],
 ['Giant Panda', '220', 'Bamboo', '20'],
 ['Red Panda', '14', 'Bamboo', '14'],
 ['Hippopotamus', '3800', 'Grass', '45'],
 ['Lion', '400', 'Gazelle', '12'],
 ['Tortoise', '100', 'Grass, Bananas', '100']]

<hr>
#### 6.4.1.3. Using `.DictReader()` to read a CSV File

Another way to interpret CSV data is as a list of dictionaries. Let's look at an example. Here is some sample CSV data:

```python
csv_data = '''Name,Weight,Diet,Average Lifespan\n
              Giraffe,2600,Shrubbery,25\n
              Moose,1300,Shrubbery,20\n
              Giant Panda,220,Bamboo,20\n'''
```   

This is the same as the following table:
                  
| Name       | Weight     | Diet      | Average Lifespan    |
| :--------- | :---------: | :------:   | :-------------------:|
| Giraffe    | 2600       | Shrubbery | 25	                |
| Moose      | 1300       | Shrubbery | 20	                |
| Giant Panda| 220        | Bamboo    | 20               	|

This same information is presented as a list of dictionaries below:

```python
csv_data = [{'Average Lifespan': '25',
             'Diet': 'Shrubbery',
             'Name': 'Giraffe',
             'Weight': '2600'},
            {'Average Lifespan': '20',
             'Diet': 'Shrubbery',
             'Name': 'Moose',
             'Weight': '1300'},
            {'Average Lifespan': '20',
             'Diet': 'Bamboo',
             'Name': 'Giant Panda',
             'Weight': '220'}]
```

Can you see the different ways each format captures the same information?

`csv.DictReader()` is similar to `csv.reader()` in function but returns a slightly different object. This object formats the data in each row as a dictionary, using the first row of data as the keys. To retain the data outside of `.DictReader()` object, append each dictionary to a list.

In [12]:
land_animals_dict = []

with open('data/animal/animal_info_land.csv') as csvfile:
    
    animal = csv.DictReader(csvfile)
    
    for row in animal:
        land_animals_dict.append(row)

land_animals_dict

[{'Average Lifespan': '25',
  'Diet': 'Shrubbery',
  'Name': 'Giraffe',
  'Weight': '2600'},
 {'Average Lifespan': '20',
  'Diet': 'Shrubbery',
  'Name': 'Moose',
  'Weight': '1300'},
 {'Average Lifespan': '20',
  'Diet': 'Bamboo',
  'Name': 'Giant Panda',
  'Weight': '220'},
 {'Average Lifespan': '14',
  'Diet': 'Bamboo',
  'Name': 'Red Panda',
  'Weight': '14'},
 {'Average Lifespan': '45',
  'Diet': 'Grass',
  'Name': 'Hippopotamus',
  'Weight': '3800'},
 {'Average Lifespan': '12',
  'Diet': 'Gazelle',
  'Name': 'Lion',
  'Weight': '400'},
 {'Average Lifespan': '100',
  'Diet': 'Grass',
  'Name': 'Tortoise',
  'Weight': '100'}]

One reason to use dictionaries over lists is that the data is more intuitively stored. Rather than having to know what order your data values are stored in, as with a list, you can call individual pieces of each row with their column name, which usually describes what the data is about. The dictionaries above look a little different from standard Python dictionaries, but they work basically the same. Below, we extract the `Name` value from the second row in our table.

In [13]:
land_animals_dict[1]['Name']

'Moose'

<hr>
#### 6.4.1.4. Using `.DictWriter()` to write to a CSV File

Writer objects can take a data collection, convert it into a string, and write it to a given file. `csv.DictWriter()` is similar to `csv.writer()`, but takes a list of dictionaries as input rather than a list of lists. When we call `.DictWriter()`, we also need to pass the fieldnames, or column names, for our data.

Again, let's alter our data slightly to demonstrate that this is really working. Below, we add a row to our dataset.

In [14]:
tiger_data = {'Name': 'Tiger', 'Diet': 'Deer', 'Average Lifespan': 10, 'Weight': 450}

land_animals_dict.append(tiger_data)
land_animals_dict

[{'Average Lifespan': '25',
  'Diet': 'Shrubbery',
  'Name': 'Giraffe',
  'Weight': '2600'},
 {'Average Lifespan': '20',
  'Diet': 'Shrubbery',
  'Name': 'Moose',
  'Weight': '1300'},
 {'Average Lifespan': '20',
  'Diet': 'Bamboo',
  'Name': 'Giant Panda',
  'Weight': '220'},
 {'Average Lifespan': '14',
  'Diet': 'Bamboo',
  'Name': 'Red Panda',
  'Weight': '14'},
 {'Average Lifespan': '45',
  'Diet': 'Grass',
  'Name': 'Hippopotamus',
  'Weight': '3800'},
 {'Average Lifespan': '12',
  'Diet': 'Gazelle',
  'Name': 'Lion',
  'Weight': '400'},
 {'Average Lifespan': '100',
  'Diet': 'Grass',
  'Name': 'Tortoise',
  'Weight': '100'},
 {'Average Lifespan': 10, 'Diet': 'Deer', 'Name': 'Tiger', 'Weight': 450}]

Now we will write the new dataset to a file. First, we specify our field names as a list of strings. Then we pass that list along with our file stream (created, as usual, with the `with open()` statement) to `.DictWriter()`. Once we have the writer object, we use it to write the table header (`.writeheader()`) and write the rows of the table (`.writerows()`).

In [15]:
with open('data/animal/animal_dict_output.csv', mode='w', newline='') as csvfile:
    
    fieldnames = ['Name', 'Weight', 'Diet', 'Average Lifespan']
    
    writer = csv.DictWriter(csvfile, fieldnames=fieldnames)
    
    writer.writeheader()
    
    writer.writerows(land_animals_dict)

Find this file in your file system and open it. Does it look how you expected it to? Let's read it back into our script to make sure it was written successfully.

In [16]:
animals_list_of_dicts = []

with open('data/animal/animal_dict_output.csv') as csvfile:
    
    animal = csv.DictReader(csvfile)
    
    for row in animal:
        animals_list_of_dicts.append(row)

animals_list_of_dicts

[{'Average Lifespan': '25',
  'Diet': 'Shrubbery',
  'Name': 'Giraffe',
  'Weight': '2600'},
 {'Average Lifespan': '20',
  'Diet': 'Shrubbery',
  'Name': 'Moose',
  'Weight': '1300'},
 {'Average Lifespan': '20',
  'Diet': 'Bamboo',
  'Name': 'Giant Panda',
  'Weight': '220'},
 {'Average Lifespan': '14',
  'Diet': 'Bamboo',
  'Name': 'Red Panda',
  'Weight': '14'},
 {'Average Lifespan': '45',
  'Diet': 'Grass',
  'Name': 'Hippopotamus',
  'Weight': '3800'},
 {'Average Lifespan': '12',
  'Diet': 'Gazelle',
  'Name': 'Lion',
  'Weight': '400'},
 {'Average Lifespan': '100',
  'Diet': 'Grass',
  'Name': 'Tortoise',
  'Weight': '100'},
 {'Average Lifespan': '10', 'Diet': 'Deer', 'Name': 'Tiger', 'Weight': '450'}]

<hr>
<a id='Glob'></a>
### 6.4.2. The Glob Library: How to Read A LOT of Data into a Python Script
<hr>
#### References:
* [Python 3.4: glob](https://docs.python.org/3.4/library/glob.html)
<hr>

You may have noticed that with each lesson we are increasing the amount of data at your fingertips. In Lesson 5 we processed a number of resume files using a list of all the file names. We were able to manually type out this list because it consisted of only ten files, but what if we want to process a thousand files? 

When typing every file name manually is out of the question, use the Glob Library. The Glob Library allows you to grab all of the file names that match a specified pattern. Yes, Glob allows you to load all of the data files in your Documents folder for example, but doing so might cause your computer to crash. It's smart to be cognizant of the amount of data you are working with and your computational resources.

The Glob Module finds all the pathnames matching a specified pattern. The syntax for implementing glob is below.

```python
files = glob.glob('folder1/folder2/*.*')
```

As always, our use of glob requires us to first import the library.

In [17]:
import glob

Now, we'll call the Glob Library's `.glob()` method. The only argument it takes is a string that supplies the pattern to match file names to. To allow `.glob()` to match files with different names, we use a wildcard character (`'*'`). The asterisk will match any amount of characters of any kind.

In [18]:
files = glob.glob('data/animal/*.*')
files

['data/animal\\animal_dict_output.csv',
 'data/animal\\animal_info_land.csv',
 'data/animal\\animal_info_sea.csv',
 'data/animal\\animal_output.csv',
 'data/animal\\bad_file.json',
 'data/animal\\bad_file.txt']

Take note of the output. We passed the `.glob()` method the pattern `'data/animal/*.*'`, which matches any file in the `data/animal/` folder with a period (`'.'`) in it. In return, we get a list of strings, each string corresponding to a file on our computer.

<b><font color="firebrick">Instructor Guidance</font>: Have the class execute the code above and discuss the questions below.</b>

#### 6.4.2.1. Discussion Questions
>1. What is the `'*'`?
>2. What type of data structure does glob return?
>3. How many filenames does `.glob()` return?
>4. What are the file types of the files that matched the pattern?
>5. How could you use the above list of file names in a script?

As we mentioned, the `'*'` is a catch-all for any string of characters. As such, the pattern `'*.*'` matches any file with a period in it. All the file names below match this pattern.

```python        
'animal.csv'
'data.json'
'info.xlsx'
```

A more specific wildcard expression would be `'*.csv'` which would only collect files with a `.csv` extension. Since we're focusing on CSV files in this lesson, let's change our previous pattern to match only CSV files.

In [19]:
files = glob.glob('data/animal/*.csv')
files

['data/animal\\animal_dict_output.csv',
 'data/animal\\animal_info_land.csv',
 'data/animal\\animal_info_sea.csv',
 'data/animal\\animal_output.csv']

NOTE: The double backslash (`'\\'`) is the way `.glob()` separates folders and files when it runs in a Windows environment. In Python, `'\\'` and `'/'` are interchangeable as path separators.

<hr>
#### 6.4.2.2. Hard-Coding vs. Programmatic Coding

Now that we're using strings to connect our Python scripts to files on our computer, let's discuss the concept of _hard-coding_. Hard-coding is when you type explicit values in your code. So far, all our paths and patterns have been hard-coded. *Programmatic coding*, on the other hand, is using variables wherever possible in your functions and method calls in lieu of explicit values.

When writing scripts for real world use cases, you need to think about how someone else could run your code on their machine.  Ask yourself: How easily can my code be applied in a slightly different context? By writing programmatic code, you make your scripts useful to others as well as yourself.

For example, let's say we write a script that approximates the diameter and area of a circle given the circle's radius. We might pick a specific radius to start out with to make sure our code works the way we want it to. Given a circle of radius 150, our code could look something like this:

In [20]:
diameter = 150 * 3.14159 * 2

area = 150**2 * 3.14159

print('diameter:', diameter)
print('area:', area)

diameter: 942.477
area: 70685.775


One thing we could do to make this code more programmatic is to store our approximation of pi (`3.14159`) in a variable, and use that variable every time we use pi in our code. Doing so would give us the following:

In [21]:
PI = 3.14159

diameter = 150 * PI * 2

area = 150**2 * PI

print('diameter:', diameter)
print('area:', area)

diameter: 942.477
area: 70685.775


Now this might not seem like a big improvement, but let's say in the future we want to make our approximation of pi more accurate. Instead of having to go in and change two values in the code (one in each of our equations) we only need to change it once, in the variable assignment, and the new value is used everywhere we call that variable. 

In [22]:
PI = 3.14159265359

diameter = 150 * PI * 2

area = 150**2 * PI

print('diameter:', diameter)
print('area:', area)

diameter: 942.477796077
area: 70685.834705775


Imagine this is part of a larger, more complicated script in which you use pi two dozen times. Coding programmatically by storing the value to a variable makes us much faster and much less error-prone in the long run.

Now our code is a little more robust thanks to our smart use of a single variable. But what if we wanted to find the diameter and area of a circle with a radius of 7500? If we're hard-coding, we'd have to go in a swap in `7500` everywhere we see the value `150` in our code above.

In [23]:
PI = 3.14159265359

diameter = 7500 * PI * 2

area = 7500**2 * PI

print('diameter:', diameter)
print('area:', area)

diameter: 47123.88980385
area: 176714586.7644375


This is doable with only two lines of code to change, but what if our script used that radius value in a hundred other places? Do you trust yourself to remember to swap every single value in your code? Better question: Why even bother if you can use a variable to do it for you?

In [24]:
PI = 3.14159265359
radius = 7500

diameter = radius * PI * 2

area = radius**2 * PI

print('diameter:', diameter)
print('area:', area)

diameter: 47123.88980385
area: 176714586.7644375


Now we can calculate the area and diameter of *any* circle just by changing a single value. Go back to the beginning of this section and notice how our code has changed. We've managed to make it more readable, more understandable, and more generalizable by using a couple of variables in the place of hard-coded values. We've also made editing and debugging easier. Coding programmatically is a good habit to build as you write your own code.

<hr>
<a id='GuidedExercise'></a>
## 6.5. Guided Exercise: Data Exploration and Engineering

<b><font color = "brick">Instructor Guidance</font>: Refer back to Lesson 1 and relate the four steps of problem-solving using Computational Thinking (Decomposition, Pattern Recognition, Abstraction, & Algorithm Design) as appropriate throughout these exercises.</b>

<hr>
#### References:
* [FlightRadar 24](https://www.flightradar24.com)
* [Wikipedia: Automatic Dependent Surveillance-Broadcast (ADS-B)](https://en.wikipedia.org/wiki/Automatic_dependent_surveillance_%E2%80%93_broadcast)
* [Wikipedia: Volga-Dnepr Airlines](https://en.wikipedia.org/wiki/Volga-Dnepr_Airlines)
* [Wikipedia: Antonov AN-124 Condor](https://en.wikipedia.org/wiki/Antonov_An-124_Ruslan)
* [JetPhotos: Volga-Dnepr AN-124 RA-82074](https://www.jetphotos.com/photo/8636024)
<hr>

This next dataset is from the online flight tracking portal FlightRadar 24, which leverages the data collected from beacons on commercial aircraft. This specific FlightRadar24 dataset is from the Volga Dnepr AN-124 Ruslan/CONDOR fleet. The filenames are the aircraft tail number/ICAO Registry followed by an underscore and a FlightRadar24 id. We're going to be cleaning and processing this data for analysis. It is always good to know what your data looks like so take a minute to open one of the data files in the `data/fr24/` folder using Notepad to look at the raw data.

Here's a sample of the data:
```python
"""
Timestamp,UTC,Callsign,Position,Altitude,Speed,Direction
1498565234,2017-06-27T12:07:14Z,VDA2164,"47.154182,-1.603088",0,18,28
1498565217,2017-06-27T12:06:57Z,VDA2164,"47.154602,-1.603609",0,22,28
1498565187,2017-06-27T12:06:27Z,VDA2164,"47.155693,-1.604902",0,29,28
1498565177,2017-06-27T12:06:17Z,VDA2164,"47.155998,-1.605566",0,28,28
1498565056,2017-06-27T12:04:16Z,VDA2164,"47.148937,-1.613806",0,24,25
"""
```

<b><font color="firebrick">Instructor Guidance</font>: The file was imported as a comma separated value file. However, sometimes Excel changes the data. So when you are troubleshooting your code, be sure to look at the raw data and not an Excel-formatted version of the data.

In the code cells below, we will work out the process for reading in and cleaning a FlightRadar 24 file.

Step 1. Read a file in using the `csv.DictReader()` method:

In [25]:
track = []

with open('data/fr24/RA64520_ccbc1f8.csv') as csvfile:
    
    d_reader = csv.DictReader(csvfile)
    
    for row in d_reader:
        track.append(row)
        
track

[{'Altitude': '2050',
  'Callsign': 'RSD73',
  'Direction': '56',
  'Position': '55.573833,37.135307',
  'Speed': '148',
  'Timestamp': '1490011088',
  'UTC': '2017-03-20T11:58:08Z'},
 {'Altitude': '2124',
  'Callsign': 'RSD73',
  'Direction': '56',
  'Position': '55.57159,37.126858',
  'Speed': '150',
  'Timestamp': '1490011058',
  'UTC': '2017-03-20T11:57:38Z'},
 {'Altitude': '2150',
  'Callsign': 'RSD73',
  'Direction': '44',
  'Position': '55.568405,37.11319',
  'Speed': '183',
  'Timestamp': '1490011049',
  'UTC': '2017-03-20T11:57:29Z'},
 {'Altitude': '2150',
  'Callsign': 'RSD73',
  'Direction': '44',
  'Position': '55.562057,37.101727',
  'Speed': '183',
  'Timestamp': '1490011043',
  'UTC': '2017-03-20T11:57:23Z'},
 {'Altitude': '2150',
  'Callsign': 'RSD73',
  'Direction': '35',
  'Position': '55.558792,37.093979',
  'Speed': '199',
  'Timestamp': '1490011023',
  'UTC': '2017-03-20T11:57:03Z'},
 {'Altitude': '2150',
  'Callsign': 'RSD73',
  'Direction': '16',
  'Position': '5

Step 2. `.DictReader()` gave us a list of dictionaries. Take a look at the first dictionary in the list and examine the data more closely.

In [26]:
track[0]

{'Altitude': '2050',
 'Callsign': 'RSD73',
 'Direction': '56',
 'Position': '55.573833,37.135307',
 'Speed': '148',
 'Timestamp': '1490011088',
 'UTC': '2017-03-20T11:58:08Z'}

It looks like the `Position` column contains strings with latitude and longitude information. This data is very useful, but hard to work with in its current form. We'd like to isolate the latitudes and longitudes and turn them into numbers.

Step 3. Split the `Position` column into two separate columns titled `'Latitude'` and `'Longitude'`, and cast the data to floats. Do this by splitting the string in the `Position` column on the comma character.

In [27]:
track[0]['Position'].split(',')

['55.573833', '37.135307']

Now let's iterate over all the rows in our data. For each row, we execute the commands to create the `Latitude` and `Longitude` columns and delete the `Position` column. Then we inspect the first row again to make sure it looks good.

In [28]:
for row in track:
    row['Latitude'] = float(row['Position'].split(',')[0])
    
    row['Longitude'] = float(row['Position'].split(',')[1])
    
    del row['Position']
    
track[0]

{'Altitude': '2050',
 'Callsign': 'RSD73',
 'Direction': '56',
 'Latitude': 55.573833,
 'Longitude': 37.135307,
 'Speed': '148',
 'Timestamp': '1490011088',
 'UTC': '2017-03-20T11:58:08Z'}

Step 4. Looks like we still have numeric data stored as strings. Format the rows `'Altitude'`, `'Speed'`, and `'Direction'` as integers.

In [29]:
for row in track:
    row['Altitude'] = int(row['Altitude'])
    
    row['Speed'] = int(row['Speed'])
    
    row['Direction'] = int(row['Direction'])
    
track[0]

{'Altitude': 2050,
 'Callsign': 'RSD73',
 'Direction': 56,
 'Latitude': 55.573833,
 'Longitude': 37.135307,
 'Speed': 148,
 'Timestamp': '1490011088',
 'UTC': '2017-03-20T11:58:08Z'}

The difference is subtle, but you can see that the quotation marks went away in the three columns we edited. This tells us we now have numeric values, not strings.

Step 5. Add a row containing the name of the file where the data came from (this will be helpful later when we look at many files with a similar structure). Let's pull the value out of the filename above, and save it into a new entry in each row. 

In [30]:
for row in track:
    row['ID'] = 'RA64520_ccbc1f8'
    
track[0]

{'Altitude': 2050,
 'Callsign': 'RSD73',
 'Direction': 56,
 'ID': 'RA64520_ccbc1f8',
 'Latitude': 55.573833,
 'Longitude': 37.135307,
 'Speed': 148,
 'Timestamp': '1490011088',
 'UTC': '2017-03-20T11:58:08Z'}

Now that we have all the steps to clean our data, let's put them all in one cell. All we need to do to read and clean a different file is change the file name we pass into the `open()` statement.

In [31]:
track = []

with open('data/fr24/RA64520_ccbc1f8.csv') as csvfile:
    d_reader = csv.DictReader(csvfile)
    for row in d_reader:
        track.append(row)

for row in track:
    row['Latitude'] = float(row['Position'].split(',')[0])
    row['Longitude'] = float(row['Position'].split(',')[1])
    del row['Position']
    row['Altitude'] = int(row['Altitude'])
    row['Speed'] = int(row['Speed'])
    row['Direction'] = int(row['Direction'])
    row['ID'] = 'RA64520_ccbc1f8'

<hr>
<a id='PracticalExercise'></a>
## 6.6. Practical Exercises

<b><font color="brick">Instructor Guidance</font>: Refer back to Lesson 1 and relate the four steps of problem-solving using Computational Thinking (Decomposition, Pattern Recognition, Abstraction, & Algorithm Design) as appropriate throughout these exercises.

<font color="brick">Instructor Guidance</font>: The practical exercises deemed most important due to content and/or a cumulative result, which should be completed first in the interest of maximum training value in relation to time are Practical Exercises 1, 2, 3, and 4. Ensure you go over the exercise solutions and (as necessary) the processes to arrive at the solutions with the students.

<font color="brick">Instructor Guidance</font>: Follow-up questions are designed to be asked by the facilitators individually as each student completes the task and has it looked at by a facilitator.</b>

<hr>
<a id='PE1'></a>
### 6.6.1. Practical Exercise 1: Create a Function to Read a File

Using the code we wrote above to create a function called `read_csv` that reads in a CSV file and returns the contents as a list of dictionaries to your script. Your functions input should be a full file name including the path, and the output should be a list of dictionaries containing all the data from the file. 

In [32]:
## YOUR CODE GOES HERE ##

In [33]:
## INSTRUCTION SOLUTION(S) ##
def read_csv(filename):
    track = []
    
    with open(filename) as csvfile:
        d_reader = csv.DictReader(csvfile)
        for row in d_reader:
            track.append(row)
    
    return track
            
read_csv('data/fr24/RA64520_ccbc1f8.csv')

[{'Altitude': '2050',
  'Callsign': 'RSD73',
  'Direction': '56',
  'Position': '55.573833,37.135307',
  'Speed': '148',
  'Timestamp': '1490011088',
  'UTC': '2017-03-20T11:58:08Z'},
 {'Altitude': '2124',
  'Callsign': 'RSD73',
  'Direction': '56',
  'Position': '55.57159,37.126858',
  'Speed': '150',
  'Timestamp': '1490011058',
  'UTC': '2017-03-20T11:57:38Z'},
 {'Altitude': '2150',
  'Callsign': 'RSD73',
  'Direction': '44',
  'Position': '55.568405,37.11319',
  'Speed': '183',
  'Timestamp': '1490011049',
  'UTC': '2017-03-20T11:57:29Z'},
 {'Altitude': '2150',
  'Callsign': 'RSD73',
  'Direction': '44',
  'Position': '55.562057,37.101727',
  'Speed': '183',
  'Timestamp': '1490011043',
  'UTC': '2017-03-20T11:57:23Z'},
 {'Altitude': '2150',
  'Callsign': 'RSD73',
  'Direction': '35',
  'Position': '55.558792,37.093979',
  'Speed': '199',
  'Timestamp': '1490011023',
  'UTC': '2017-03-20T11:57:03Z'},
 {'Altitude': '2150',
  'Callsign': 'RSD73',
  'Direction': '16',
  'Position': '5

<hr>
<a id='PE2'></a>
### 6.6.2. Practical Exercise 2: Clean a CSV File

Now add the code we developed in the Guided Exercise to clean the flight radar data to your `read_csv` function. When you're done, your function should read, clean, and return the data to your script.

Cleaning Steps:
- Split the `'Position'` column into two separate columns for `'Latitude'` and `'Longitude.'`
- Format the data in `'Altitude'`, `'Speed'`, and `'Direction'` as integers.
- Add an ID to the data that indicates what file each data point came from. Do this using string operations on the filename.

BONUS: Write a [docstring](https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings) explaining what your function does, its input(s), and its output(s).

In [34]:
## YOUR CODE GOES HERE ##

In [35]:
## INSTRUCTION SOLUTION(S) ##

def read_csv(filename):
    """Reads in and cleans FR_24 data.
    
    Args:
        filename -- (str) a file name including the relative path to the file
        
    Returns:
        track -- (list of dict) the cleaned data
    """
    
    track = []
    
    with open(filename) as csvfile:
        reader = csv.DictReader(csvfile)
        for row in reader:
            track.append(row)
            
    for row in track:
        row['Latitude'] = float(row['Position'].split(',')[0])
        row['Longitude'] = float(row['Position'].split(',')[1])
        del row['Position']
        row['Altitude'] = int(row['Altitude'])
        row['Speed'] = int(row['Speed'])
        row['Direction'] = int(row['Direction'])
        row['ID'] = filename[10:-4]
        
    return track


read_csv('data/fr24/RA64520_ccbc1f8.csv')

[{'Altitude': 2050,
  'Callsign': 'RSD73',
  'Direction': 56,
  'ID': 'RA64520_ccbc1f8',
  'Latitude': 55.573833,
  'Longitude': 37.135307,
  'Speed': 148,
  'Timestamp': '1490011088',
  'UTC': '2017-03-20T11:58:08Z'},
 {'Altitude': 2124,
  'Callsign': 'RSD73',
  'Direction': 56,
  'ID': 'RA64520_ccbc1f8',
  'Latitude': 55.57159,
  'Longitude': 37.126858,
  'Speed': 150,
  'Timestamp': '1490011058',
  'UTC': '2017-03-20T11:57:38Z'},
 {'Altitude': 2150,
  'Callsign': 'RSD73',
  'Direction': 44,
  'ID': 'RA64520_ccbc1f8',
  'Latitude': 55.568405,
  'Longitude': 37.11319,
  'Speed': 183,
  'Timestamp': '1490011049',
  'UTC': '2017-03-20T11:57:29Z'},
 {'Altitude': 2150,
  'Callsign': 'RSD73',
  'Direction': 44,
  'ID': 'RA64520_ccbc1f8',
  'Latitude': 55.562057,
  'Longitude': 37.101727,
  'Speed': 183,
  'Timestamp': '1490011043',
  'UTC': '2017-03-20T11:57:23Z'},
 {'Altitude': 2150,
  'Callsign': 'RSD73',
  'Direction': 35,
  'ID': 'RA64520_ccbc1f8',
  'Latitude': 55.558792,
  'Longitude

<hr>
<a id='PE3'></a>
### 6.6.3. Practical Exercise 3: Read in a CSV File Using the Glob Library
Use your function to read in all of the CSV files found in the `fr24/` folder. Save all of the data to one data structure. Do not include files that have no information. Find the length of your new data structure.

HINT: use the Glob Library to get all the filenames for CSVs in the `fr24/` folder.

WARNING: This is a lot of data, so printing it to your screen can crash your notebook.

In [36]:
## YOUR CODE GOES HERE ##

In [37]:
## INSTRUCTION SOLUTION(S) ##
files = glob.glob('data/fr24/*.csv')

flighttracks = []

for filename in files:
    
    track = read_csv(filename)
        
    if len(track) > 0:
        flighttracks.append(track)
    
len(flighttracks)

1855

Part 2: What type of data structure is the one you just created? Use the `type()` function to find out.

In [38]:
## YOUR CODE GOES HERE ##

In [39]:
## INSTRUCTION SOLUTION(S) ##
print(type(flighttracks))

print(type(flighttracks[0]))

print(type(flighttracks[0][0]))

<class 'list'>
<class 'list'>
<class 'dict'>


<hr>
<a id='PE4'></a>
### 6.6.4. Practical Exercise 4: Make a Dictionary
Use the data structure you made in Practical Exercise 3 to answer the following questions.

Part 1: Each CSV your read in above contains data for a single flight. In that data, the callsign value identifies the plane. How would you access the callsign for the first flight in your big list of flights?

In [40]:
## YOUR CODE GOES HERE ##

In [41]:
## INSTRUCTION SOLUTION(S) ##

flighttracks[0][0]['Callsign']

'BOE004'

Part 2: Make a dictionary that uses the `'Callsign'` values as keys. For each callsign, the associated value in your new dictionary should be the number of rows of data that contain that callsign.

HINT: Use a `for` loop to iterate through each flight, adding a new dictionary entry for each new callsign you find.

In [42]:
## YOUR CODE GOES HERE ##

In [43]:
## INSTRUCTION SOLUTION(S) ##
callsigns = {}
    
for track in flighttracks:
    
    current_callsign = track[0]['Callsign']
    
    if current_callsign in callsigns:
        callsigns[current_callsign] += len(track)
        
    else:
        callsigns[current_callsign] = len(track)

callsigns

{'': 10915,
 '38416': 497,
 '79941': 130,
 'ADB200F': 642,
 'ADB204F': 305,
 'ADB205F': 125,
 'ADB2092': 910,
 'ADB2100': 655,
 'ADB2104': 549,
 'ADB2105': 697,
 'ADB2110': 1481,
 'ADB2127': 393,
 'ADB2130': 251,
 'ADB2131': 206,
 'ADB2132': 349,
 'ADB2133': 1357,
 'ADB2134': 2163,
 'ADB2135': 2219,
 'ADB2136': 731,
 'ADB2137': 290,
 'ADB2138': 974,
 'ADB2139': 415,
 'ADB2142': 285,
 'ADB2143': 399,
 'ADB2146': 953,
 'ADB2147': 461,
 'ADB2148': 697,
 'ADB2149': 728,
 'ADB2152': 769,
 'ADB2153': 1797,
 'ADB2154': 479,
 'ADB2155': 380,
 'ADB2158': 1238,
 'ADB2160': 273,
 'ADB2161': 201,
 'ADB2167': 1419,
 'ADB2171': 547,
 'ADB2173': 563,
 'ADB2174': 1223,
 'ADB2175': 1266,
 'ADB2176': 557,
 'ADB2177': 1073,
 'ADB2178': 1154,
 'ADB2179': 485,
 'ADB2180': 125,
 'ADB2183': 777,
 'ADB2184': 30,
 'ADB2185': 993,
 'ADB2186': 425,
 'ADB2187': 331,
 'ADB2188': 309,
 'ADB2189': 178,
 'ADB2191': 1037,
 'ADB2193': 149,
 'ADB227F': 556,
 'ADB230F': 198,
 'ADB231F': 214,
 'ADB233F': 382,
 'ADB234F': 

Part 3: Find the call signs that have more than 800 rows.

In [44]:
## YOUR CODE GOES HERE ##

In [45]:
## INSTRUCTION SOLUTION(S) ##
for callsign, num_rows in callsigns.items():
    if num_rows > 800:
        print(callsign)


ADB247F
RSD091
VDA2290
MXM3723
VDA2291
RSD076
VDA2075
RSD980
ADB2178
MXM3724
RSD137
VDA2632
VDA2138
RSD111
VDA2820
VDA8316
ADB242F
VDA2212
ADB234F
ADB2185
RSD039
ADB2153
VDA6604
RSD072
VDA2543
ADB2174
VDA6954
ADB289F
VDA7159
VDA2496
VDA4992
VDA3026
VDA3073
VDA2234
VDA2494
RSD038
RA82013
MXM3725
ADB380F
RSD048
VDA8311
ADB5593
ADB3865
RSD074
VDA4008
VDA2852
RFF8065
VDA2689
VDA4910
VDA2213
ADB2191
VDA2057
ADB2158
VDA4756
VDA9042
VDA4917
VDA8299
VDA2680
MXM3722
VDA2647
RSD47
ADB5574
VDA9686
RSD032
ADB2135
VDA2253
RSD135
ADB5579
RSD047
VDA4814
RA82010
RA82014
VDA9110
VDA9119
RSD075
VDA4886
VDA7974
VDA5012
VDA8187
VDA2118
RSD010
ADB5600
VDA5137
ADB263F
VDA6557
VDA7158
ADB2167
VDA7116
VDA7010
VDA9116
VDA8194
BOE004
ADB5578
VDA8298
ADB2133
ADB2110
VDA2252
RSD073
VDA2056
VDA2185
VDA2436
RSD280
VDA4904
VDA2208
VDA2626
VDA2586
RSD264
RSD033
ADB2175
VDA5143
ADB248F
ADB576F
VDA3024
RA82035
VDA4916
VDA2575
VDA7038
VDA9044
RSD077
VDA2209
ADB3880
VDA4884
VDA2254
VDA2851
RSD031
VDA2574
VDA7980
ADB2138

Part 4: Save the call signs with more than 800 rows into a list. For each callsign, save the callsign and its number of rows as a two item list, like this:

```python
['RSD711', 1247]
```

In [46]:
## YOUR CODE GOES HERE ##

In [47]:
## INSTRUCTION SOLUTION(S) ##
over_800 = []

for callsign, num_rows in callsigns.items():
    if num_rows > 800:
        over_800.append([callsign, num_rows])
        
over_800

[['', 10915],
 ['ADB247F', 802],
 ['RSD091', 1319],
 ['VDA2290', 850],
 ['MXM3723', 2201],
 ['VDA2291', 930],
 ['RSD076', 2703],
 ['VDA2075', 919],
 ['RSD980', 3352],
 ['ADB2178', 1154],
 ['MXM3724', 1767],
 ['RSD137', 915],
 ['VDA2632', 1536],
 ['VDA2138', 1541],
 ['RSD111', 2023],
 ['VDA2820', 866],
 ['VDA8316', 1048],
 ['ADB242F', 1078],
 ['VDA2212', 2515],
 ['ADB234F', 1234],
 ['ADB2185', 993],
 ['RSD039', 1100],
 ['ADB2153', 1797],
 ['VDA6604', 1553],
 ['RSD072', 1068],
 ['VDA2543', 1272],
 ['ADB2174', 1223],
 ['VDA6954', 968],
 ['ADB289F', 962],
 ['VDA7159', 1052],
 ['VDA2496', 1429],
 ['VDA4992', 888],
 ['VDA3026', 953],
 ['VDA3073', 975],
 ['VDA2234', 1174],
 ['VDA2494', 951],
 ['RSD038', 2056],
 ['RA82013', 2655],
 ['MXM3725', 1234],
 ['ADB380F', 917],
 ['RSD048', 2211],
 ['VDA8311', 1086],
 ['ADB5593', 1011],
 ['ADB3865', 1445],
 ['RSD074', 1790],
 ['VDA4008', 1259],
 ['VDA2852', 1321],
 ['RFF8065', 1232],
 ['VDA2689', 959],
 ['VDA4910', 883],
 ['VDA2213', 868],
 ['ADB2191', 

Part 5: Write the new structure you just created to a CSV file at the following location:

```python
'data/well_used_callsigns.csv'
```

In [48]:
## YOUR CODE GOES HERE ##

In [49]:
## INSTRUCTION SOLUTION(S) ##
with open('data/well_used_callsigns.csv', 'w', newline='') as csvfile:
    
    writer = csv.writer(csvfile)
    
    writer.writerows(over_800)

<hr>
<a id='PE5'></a>
### 6.6.5. Practical Exercise 5: Read a Sample Dataset

Use the CSV Library to read a sample personnel dataset found at the following location:
```python
'data/random_personnel.csv'
```
Count all of the area codes found in the `'Phone Number'` column (2nd Column).

Phone Number format: (XXX) XXX-XXXX

HINT: Create a dictionary that will have the area codes as keys and their respective counts as values. Add/increment the count for area codes as you iterate through the CSV file.

In [50]:
## YOUR CODE GOES HERE ##

In [51]:
## INSTRUCTION SOLUTION(S) ##
ac_counts = {}

with open('data/random_personnel.csv') as csvfile:
    
    reader = csv.reader(csvfile)
    
    data = []
    
    for row in reader:
        data.append(row)
        
    for row in data[1:]:  ## the first row is the header
        phone_num = row[1]
        area_code = phone_num[1:4]
        if area_code in ac_counts:
            ac_counts[area_code] += 1
        else: 
            ac_counts[area_code] = 1

ac_counts

{'106': 1,
 '107': 1,
 '111': 1,
 '128': 2,
 '135': 1,
 '142': 1,
 '145': 1,
 '150': 1,
 '159': 1,
 '169': 1,
 '178': 1,
 '265': 1,
 '274': 1,
 '287': 1,
 '292': 1,
 '324': 1,
 '325': 1,
 '335': 1,
 '337': 1,
 '351': 1,
 '366': 1,
 '368': 1,
 '379': 2,
 '390': 1,
 '396': 1,
 '398': 1,
 '410': 1,
 '416': 1,
 '423': 1,
 '436': 1,
 '437': 1,
 '439': 1,
 '460': 1,
 '471': 1,
 '476': 1,
 '479': 1,
 '482': 1,
 '491': 1,
 '497': 1,
 '498': 1,
 '539': 1,
 '541': 1,
 '547': 1,
 '558': 1,
 '566': 1,
 '580': 1,
 '583': 1,
 '590': 1,
 '597': 1,
 '598': 1,
 '603': 1,
 '608': 2,
 '611': 1,
 '619': 1,
 '631': 1,
 '641': 1,
 '645': 1,
 '666': 1,
 '683': 1,
 '703': 1,
 '704': 1,
 '719': 1,
 '723': 1,
 '737': 1,
 '740': 1,
 '741': 1,
 '743': 2,
 '748': 1,
 '751': 1,
 '762': 1,
 '777': 1,
 '790': 1,
 '791': 1,
 '810': 2,
 '812': 1,
 '844': 1,
 '845': 1,
 '847': 1,
 '849': 1,
 '858': 1,
 '868': 2,
 '878': 1,
 '903': 2,
 '907': 2,
 '919': 1,
 '920': 1,
 '922': 1,
 '948': 1,
 '980': 1,
 '991': 2}

<hr>
<a id='Administrative'></a>
## 6.7. Administrative Notes:
* Save your Lesson 6 .ipynb file to your H drive

<b><center>UNCLASSIFIED</center></b>